NameLayerMain/src/vg/civcraft/mc/namelayer/GroupManager.java

```diff
@@ -330,25 +330,21 @@ public boolean hasAccess(Group group, UUID player, PermissionType perm) {
 			}
 		}
 		GroupPermission perms = getPermissionforGroup(group);
-		for(PlayerType rank : getRecursivePlayerTypes(group, player)) {
-			if (perms.hasPermission(rank, perm)) {
-				//player has right rank in the group itself or at least one super group
-				return true;
-			}
-		}		
+		PlayerType rank = getRecursiveHighestPlayerType(group, player); 
+		if (perms.hasPermission(rank, perm)) {
+			//player has right rank in the group itself or at least one super group
+			return true;
+		}	
 		return false;
 	}
 	
-	private List<PlayerType> getRecursivePlayerTypes(Group group, UUID player) {
-		List<PlayerType> perms = new LinkedList<PlayerType>();
+	private PlayerType getRecursiveHighestPlayerType(Group group, UUID player) {
 		PlayerType type = group.getPlayerType(player);
-		if (type != null) {
-			perms.add(type);
-		}
 		if (group.hasSuperGroup()) {
-			perms.addAll(getRecursivePlayerTypes(group.getSuperGroup(), player));
+			//only return highest rank
+			type = PlayerType.getByID(Math.max(PlayerType.getID(type), PlayerType.getID(getRecursiveHighestPlayerType(group.getSuperGroup(), player))));
 		}
-		return perms;
+		return type;
```

Comment:
With respect to the permission recursion change:

Hmm, I'm not clear what benefit this refactor provides. It might be potentially a bit faster, but it's worth noting that using getByID on the enum is not a good practice -- future types might be injected in the middle by programmers unaware of this structured code, leading to very hard to debug regressions. 

The other issue I can imagine -- although it's not normal practice -- is that may unintentionally break "odd" permission sets within subgroup structures. By way of example, perhaps MODS on a group are given DOORS, but not CHESTS; MEMBERS are given CHESTS but not DOORS. A player has MODS on the supergroup but MEMBER on the subgroup. If you always assume that the higher permission _type_ equates to equivalence and expansion of lower type permissions, this creates an opening for weird scenarios like this. The old code would have worked fine with this configuration; the new code will only do the check against MODS (the highest permtype within the linked groups) and never grant CHESTS access, even though the player has proper access (MEMBERS on the subgroup).